wallet/src/key_chain/leaf_key_chain.rs

```diff
@@ -0,0 +1,517 @@
+            purpose,
+            parent_pubkey,
+            addresses: BTreeMap::new(),
+            derived_public_keys: BTreeMap::new(),
+            public_key_to_index: BTreeMap::new(),
+            public_key_hash_to_index: BTreeMap::new(),
+            usage_state: KeychainUsageState::default(),
+        }
+    }
+
+    // TODO reduce the number of parameters
+    #[allow(clippy::too_many_arguments)]
+    fn new_from_parts(
+        chain_config: Arc<ChainConfig>,
+        account_id: AccountId,
+        purpose: KeyPurpose,
+        parent_pubkey: ExtendedPublicKey,
+        addresses: BTreeMap<ChildNumber, Address>,
+        derived_public_keys: BTreeMap<ChildNumber, ExtendedPublicKey>,
+        usage_state: KeychainUsageState,
+    ) -> KeyChainResult<Self> {
+        // TODO optimize for database structure
+        let public_keys_to_index: BTreeMap<PublicKey, ChildNumber> = derived_public_keys
+            .iter()
+            .map(|(idx, xpub)| (xpub.clone().into_public_key(), *idx))
+            .collect();
+        let public_key_hashes_to_index: BTreeMap<PublicKeyHash, ChildNumber> = public_keys_to_index
+            .iter()
+            .map(|(pk, idx)| (PublicKeyHash::from(pk), *idx))
+            .collect();
+
+        Ok(Self {
+            chain_config,
+            account_id,
+            purpose,
+            parent_pubkey,
+            addresses,
+            derived_public_keys,
+            public_key_to_index: public_keys_to_index,
+            public_key_hash_to_index: public_key_hashes_to_index,
+            usage_state,
+        })
+    }
+
+    pub fn load_leaf_keys<B: Backend>(
+        chain_config: Arc<ChainConfig>,
+        account_info: &DeterministicAccountInfo,
+        db_tx: &StoreTxRo<B>,
+        id: &AccountId,
+    ) -> KeyChainResult<(LeafKeyChain, LeafKeyChain)> {
+        let mut receiving_addresses = BTreeMap::new();
+        let mut change_addresses = BTreeMap::new();
+
+        for (address_id, address) in db_tx.get_addresses(id)? {
+            let (purpose, key_index) = get_purpose_and_index(&address_id.into_item_id())?;
+            let old_value = match purpose {
+                KeyPurpose::ReceiveFunds => receiving_addresses.insert(key_index, address),
+                KeyPurpose::Change => change_addresses.insert(key_index, address),
+            };
+            if old_value.is_some() {
+                return Err(KeyChainError::CouldNotLoadKeyChain);
+            }
+        }
+
+        let mut receiving_public_keys = BTreeMap::new();
+        let mut change_public_keys = BTreeMap::new();
+        for (pubkey_id, xpub) in db_tx.get_public_keys(id)? {
+            let (purpose, key_index) = get_purpose_and_index(&pubkey_id.into_item_id())?;
+            let old_value = match purpose {
+                KeyPurpose::ReceiveFunds => receiving_public_keys.insert(key_index, xpub),
+                KeyPurpose::Change => change_public_keys.insert(key_index, xpub),
+            };
```

Comment:
Here's a random idea, feel free to run with it, modify or ignore as you please.

I noticed you often have a pairs of things (here `BTreeMap` with public keys) where one of the pairs is used with `ReceiveFunds` purpose and the other is used for `Change` purpose. You could define a helper container that packages these together and gives you convenient access to the object corresponding to given purpose without having to `match` every time.

Rough outline:
```rust
struct WithPurpose<T> { // TODO: name?
    receive: T,
    change: T,
}

impl<T> WithPurpose<T> {
    // implementations obvious, omitted
    fn new(receive: T, change: T) -> Self;
    fn ref_for(&self, purpose: KeyPurpose) -> &T;
    fn mut_for(&mut self, purpose: KeyPurpose) -> &mut T;
    // Optional, useful combined with `flat_map`:
    fn into_iter(&mut self) -> impl Iterator<Item = T> + ExactSizeIterator + '_;
    fn iter(self) -> impl Iterator<Item = &T> + ExactSizeIterator;
    fn iter_mut(&mut self) -> impl Iterator<Item = &mut T> + ExactSizeIterator + '_;
}
```

The above code could then be something like this, given `public_keys: WithPurpose<BTreeMap<_, _>>`:
```rust
let old_value = public_keys.mut_for(purpose).insert(key_index, xpub);
```

Also the return value of this whole function (`load_leaf_keys`) could be `WithPurpose<LeafKeyChain>` instead of `(LeafKeyChain, LeafKeyChain)`.